### PR TITLE
[Feature] remove implicit cache function in DDF

### PR DESCRIPTION
**Pull request by Huandao0812**
### Description and related tickets, documents

Add DDF api to cache, uncache a DDF.
if a DDF is cached, it'll also be cached after an in-place transformation.
new DDF will always be uncached.

Reviewers: 
- Main reviewer: @ubolonton 
- Optional reviewer: @lebinh @zkidkid 
- Observers: @hai-adatao
### PR Progress

Make sure all checkboxes below are checked before merged
- [x] Merge check has no conflicts. PR checks passed.
- [x] Main reviewer approved
- [x] Optional reviewer approved

### CODE REVIEW

File: spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package io.ddf.spark
+
+import io.ddf.spark.content.RepresentationHandler
+
+class CacheDDFSuite extends ATestSuite {
+  createTableAirline()
+
+  test("cache a ddf") {
+    val ddf = manager.sql2ddf("select * from airline", "SparkSQL")
```
Comment (nqbao) [inline spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala:8]: does this mean that the DDF created from SQL is always uncached?

File: spark/src/main/java/io/ddf/spark/content/SchemaHandler.java [hunk 2]
```diff
@@ -102,7 +102,11 @@ public static Schema getSchemaFromDataFrame(DataFrame rdd) {
 
       for (Integer columnIndex : columnIndexes) {
         String colName = this.getDDF().getColumnName(columnIndex);
-        listLevelCountsWithName.put(colName, listLevelCounts.get(columnIndex));
+        if(listLevelCounts.get(columnIndex) != null) {
```
Comment (nqbao) [inline spark/src/main/java/io/ddf/spark/content/SchemaHandler.java]: is this check because when ddf is not cached, the factors are not available?

Comment (Huandao0812) [inline spark/src/main/java/io/ddf/spark/content/SchemaHandler.java]: just handle a case where the levels count is empty for the current column, it's a small bug not related to this PR

File: core/src/main/java/io/ddf/DDF.java [hunk 3]
```diff
@@ -83,6 +85,20 @@ public void setIsDDFView(boolean isDDFView) {
     this.mIsDDFView = isDDFView;
   }
 
+  public boolean isCache() {
+    return this.isCache;
+  }
+
+  public void cache() {
+    this.isCache = true;
+    this.getRepresentationHandler().cache(false);
```
Comment (lebinh) [inline core/src/main/java/io/ddf/DDF.java]: Better be?

```
if (!this.isCached()) {
  // do caching
}
```

so we have better control and callers don't have to check

File: core/src/main/java/io/ddf/DDF.java [hunk 4]
```diff
@@ -83,6 +85,20 @@ public void setIsDDFView(boolean isDDFView) {
     this.mIsDDFView = isDDFView;
   }
 
+  public boolean isCache() {
```
Comment (lebinh) [inline core/src/main/java/io/ddf/DDF.java]: Should be named `isCached`

File: core/src/main/java/io/ddf/DDF.java (hunk 3 above)
Comment (ubolonton) [inline core/src/main/java/io/ddf/DDF.java]: What happens when this throw an exception? How about when it's called by multiple code paths? Maybe there should be a lock on `isCached`?

Comment (hai-adatao) [inline core/src/main/java/io/ddf/DDF.java]: Also, in `cache` and `uncache` methods, the flag `isCached` always changes the value upon calling, should only change this after the operation finishes successfully?

Comment (Huandao0812) [inline core/src/main/java/io/ddf/DDF.java]: yes changing that

File: core/src/main/java/io/ddf/content/MutabilityHandler.java [hunk 5]
```diff
@@ -28,11 +28,14 @@ public void setMutable(boolean isMutable) {
   public DDF updateInplace(DDF newddf) throws DDFException {
     //copy content of newddf to this ddf
     DDF curDDF = this.getDDF();
+    //cache the new representation if the current ddf is cached
+    if(curDDF.isCache()) {
```
Comment (hai-adatao) [inline core/src/main/java/io/ddf/content/MutabilityHandler.java]: Once we cache new DDF, do we need to uncache old DDF to claim some memory?

File: core/src/main/java/io/ddf/content/MutabilityHandler.java [hunk 6]
```diff
@@ -28,11 +28,14 @@ public void setMutable(boolean isMutable) {
...
+      newddf.cache();
+    }
     curDDF.getRepresentationHandler().reset();
```
Comment (Huandao0812) [inline core/src/main/java/io/ddf/content/MutabilityHandler.java:35]: @hai-adatao this line will do the uncaching the previous DDF

File: spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package io.ddf.spark
+
+
+class CacheDDFSuite extends ATestSuite {
+  createTableAirline()
+
+  test("cache a ddf") {
+    val ddf = manager.sql2ddf("select * from airline", "SparkSQL")
+    ddf.getRepresentationHandler.cache(false)
```
Comment (hai-adatao) [inline spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala:9]: Can we assert that the DDF is created uncached before running this?

Comment (Huandao0812) [inline spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala:9]: this is checked inside representationHandler.cache()

File: spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    assert(ddf.getRepresentationHandler.isCached)
+
+    ddf.Transform.transformUDFWithNames(Array("newcol"), Array("arrdelay + depdelay"), Array("arrdelay", "depdelay"),
```
Comment (hai-adatao) [inline spark/src/test/scala/io/ddf/spark/CacheDDFSuite.scala:12]: Can we change to `ddfr = ddf.Transform...` and verify that both `ddfr` and `ddf` are cached? Similar for the test below.

**Comment by Huandao0812** [comment]
@nhanitvn I changed summy and getFactorLevelCounts in this PR, please help review

File: spark/src/main/java/io/ddf/spark/analytics/BasicStatisticsComputer.java [hunk 9]
```diff
@@ -37,11 +37,36 @@ public BasicStatisticsComputer(DDF theDDF) {
...
+    int i = 1;
+    for(Column column: columns) {
+      if(column.isNumeric()) {
+        long count = Long.valueOf(rows[0].getString(i));
+        double mean = Double.valueOf(rows[1].getString(i));
+        double stddev = Double.valueOf(rows[2].getString(i));
+        double sqdeviation = Math.pow(stddev, 2)*(count - 1);
+        double min = Double.valueOf(rows[3].getString(i));
+        double max = Double.valueOf(rows[4].getString(i));
+        String sqlCmd = String.format("select count(*) from %s where `%s` is null",
```
Comment (ubolonton) [inline spark/src/main/java/io/ddf/spark/analytics/BasicStatisticsComputer.java:80]: Wouldn't it be faster if we use `sum(if(...))` in a single query?

Comment (Huandao0812) [inline spark/src/main/java/io/ddf/spark/analytics/BasicStatisticsComputer.java:80]: no bandwidth for that

**Comment by lebinh** [comment]
lgtm

**Pull request merged**